File: lua/lockfile/parse/yarn.lua
--- yarn.lock -> normalized model.
---
--- Two on-disk formats share this filename:
---   * Yarn Classic (v1): a bespoke indentation format ("name@range":\n  version
---     "x"\n  dependencies:\n    dep "range"). Parsed by hand below.
---   * Yarn Berry (v2+): valid YAML with a `__metadata` block. Parsed via the
---     YAML parser.

local yaml = require("lockfile.parse.yaml")
local model = require("lockfile.model")
local util = require("lockfile.util")

local M = {}

--- Extract the package name from a descriptor like "name@range",
--- "@scope/name@range", or Berry's "name@npm:range".
---@param descriptor string
---@return string
local function name_from_descriptor(descriptor)
  local s = util.trim(descriptor)
  -- strip surrounding quotes
  if (s:sub(1, 1) == '"' and s:sub(-1) == '"') or (s:sub(1, 1) == "'" and s:sub(-1) == "'") then
    s = s:sub(2, -2)
  end
  -- name is everything before the last '@' that is past a leading scope '@'.
  for i = #s, 2, -1 do
    if s:sub(i, i) == "@" then
      return s:sub(1, i - 1)
    end
  end
  return s
end

--- Strip surrounding double/single quotes from a value, if present.
---@param s string
---@return string
local function unquote(s)
  s = util.trim(s)
  if #s >= 2 and ((s:sub(1, 1) == '"' and s:sub(-1) == '"') or (s:sub(1, 1) == "'" and s:sub(-1) == "'")) then
    return s:sub(2, -2)
  end
  return s
end

--- Split a Yarn Classic header line into individual descriptors. Descriptors
--- are separated by ", "; quoted descriptors are handled so commas inside
--- (rare) are not split points.
---@param header string
---@return string[]
local function split_descriptors(header)
  local out = {}
  local buf = {}
  local in_quote = nil ---@type string?
  local i = 1
  while i <= #header do
    local c = header:sub(i, i)
    if in_quote then
      buf[#buf + 1] = c
      if c == in_quote then
        in_quote = nil
      end
    elseif c == '"' or c == "'" then
      in_quote = c
      buf[#buf + 1] = c
    elseif c == "," then
      -- consume following space(s)
      out[#out + 1] = util.trim(table.concat(buf))
      buf = {}
      while i + 1 <= #header and header:sub(i + 1, i + 1) == " " do
        i = i + 1
      end
    else
      buf[#buf + 1] = c
    end
    i = i + 1
  end
  if #buf > 0 then
    out[#out + 1] = util.trim(table.concat(buf))
  end
  return out
end

--- Build logical (indent, content) lines, dropping comments and blank lines.
---@param src string
---@return {indent: integer, content: string}[]
local function logical_lines(src)
  local lines = {}
  local start = 1
  local len = #src
  for i = 1, len + 1 do
    local c = (i <= len) and src:sub(i, i) or "\n"
    if c == "\n" then
      local raw = src:sub(start, i - 1)
      if raw:sub(-1) == "\r" then
        raw = raw:sub(1, -2)
      end
      local indent = 0
      while indent < #raw and raw:sub(indent + 1, indent + 1) == " " do
        indent = indent + 1
      end
      local content = util.trim(raw:sub(indent + 1))
      if content ~= "" and content:sub(1, 1) ~= "#" then
        lines[#lines + 1] = { indent = indent, content = content }
      end
      start = i + 1
    end
  end
  return lines
end

--- Split a field line "key value" into key and (unquoted) value.
---@param content string
---@return string key
---@return string value
local function split_field(content)
  local sp = content:find(" ", 1, true)
  if not sp then
    return content, ""
  end
  return content:sub(1, sp - 1), unquote(content:sub(sp + 1))
end

--- Parse Yarn Classic (v1) format.
---@param src string
---@return lockfile.Lockfile
local function build_classic(src)
  local lf = model.new("yarn")
  lf.format_version = "1"
  local lines = logical_lines(src)
  local n = #lines
  local i = 1
  while i <= n do
    local line = lines[i]
    if line.indent == 0 then
      -- header: descriptors..., trailing ':'
      local header = line.content
      if header:sub(-1) == ":" then
        header = header:sub(1, -2)
      end
      local descriptors = split_descriptors(header)
      i = i + 1

      local version, resolved, integrity
      local deps = {}
      while i <= n and lines[i].indent > 0 do
        local f = lines[i]
        if f.content:sub(-1) == ":" then
          -- sub-block, e.g. dependencies:
          local block_key = util.trim(f.content:sub(1, -2))
          local block_indent = f.indent
          i = i + 1
          while i <= n and lines[i].indent > block_indent do
            if block_key == "dependencies" or block_key == "optionalDependencies" then
              local dep_name = unquote((split_field(lines[i].content)))
              if dep_name ~= "" then
                deps[#deps + 1] = { name = dep_name }
              end
            end
            i = i + 1
          end
        else
          local key, value = split_field(f.content)
          if key == "version" then
            version = value
          elseif key == "resolved" then
            resolved = value
          elseif key == "integrity" then
            integrity = value
          end
          i = i + 1
        end
      end

      if version and descriptors[1] then
        local name = name_from_descriptor(descriptors[1])
        lf:add({
          name = name,
          version = version,
          source = resolved,
          checksum = integrity,
          deps = deps,
        })
      end
    else
      i = i + 1
    end
  end
  return lf
end

--- Parse Yarn Berry (v2+) YAML format.
---@param src string
---@return lockfile.Lockfile
local function build_berry(src)
  local data = yaml.parse(src)
  local lf = model.new("yarn")
  if type(data) ~= "table" then
    return lf
  end
  if type(data.__metadata) == "table" and data.__metadata.version ~= nil then
    lf.format_version = tostring(data.__metadata.version)
  end

  for key, entry in pairs(data) do
    if key ~= "__metadata" and type(key) == "string" and type(entry) == "table" then
      local version = entry.version
      if version ~= nil and version ~= yaml.NULL then
        local first = split_descriptors(key)[1] or key
        local name = name_from_descriptor(first)
        local deps = {}
        if type(entry.dependencies) == "table" then
          for dname in pairs(entry.dependencies) do
            if type(dname) == "string" then
              deps[#deps + 1] = { name = dname }
            end
          end
        end
        local source
        if type(entry.resolution) == "string" then
          source = entry.resolution
        end
        local checksum
        if type(entry.checksum) == "string" then
          checksum = entry.checksum
        end
        lf:add({
          name = name,
          version = tostring(version),
          source = source,
          checksum = checksum,
          deps = deps,
        })
      end
    end
  end
  return lf
end

--- Build a normalized lockfile from yarn.lock source text, auto-detecting the
--- Classic vs Berry format.
---@param src string
---@return lockfile.Lockfile
function M.build(src)
  -- Berry lockfiles always contain a `__metadata` block and are valid YAML.
  if src:find("__metadata", 1, true) then
    return build_berry(src)
  end
  return build_classic(src)
end

return M

File: src/parse/yarn.rs
//! yarn.lock -> normalized model.
//!
//! Two formats share this filename: Yarn Classic (v1, a bespoke indentation
//! format) and Yarn Berry (v2+, valid YAML with a `__metadata` block). They are
//! distinguished by the presence of `__metadata` and dispatched accordingly.

use serde_yaml_ng::Value;

use super::yarn_classic;
use crate::model::{Dep, Lockfile, Package};

pub fn parse(src: &str) -> Result<Lockfile, String> {
    if src.contains("__metadata") {
        parse_berry(src)
    } else {
        yarn_classic::parse(src)
    }
}

fn value_to_string(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        Value::Number(n) => n.to_string(),
        Value::Bool(b) => b.to_string(),
        _ => String::new(),
    }
}

fn parse_berry(src: &str) -> Result<Lockfile, String> {
    let doc: Value = serde_yaml_ng::from_str(src).map_err(|e| format!("invalid yarn.lock: {e}"))?;
    let map = doc
        .as_mapping()
        .ok_or_else(|| "invalid yarn.lock: expected a mapping at the top level".to_string())?;

    let mut lf = Lockfile::new("yarn");
    if let Some(meta) = map.get("__metadata").and_then(Value::as_mapping) {
        if let Some(version) = meta.get("version") {
            lf.format_version = Some(value_to_string(version));
        }
    }

    for (key, entry) in map {
        let Some(key) = key.as_str() else { continue };
        if key == "__metadata" {
            continue;
        }
        let Some(entry) = entry.as_mapping() else {
            continue;
        };
        let Some(version) = entry.get("version").and_then(Value::as_str) else {
            continue;
        };

        let first = key.split(", ").next().unwrap_or(key);
        let name = yarn_classic::name_from_descriptor(first);
        let mut pkg = Package::new(name, version);
        pkg.source = entry
            .get("resolution")
            .and_then(Value::as_str)
            .map(str::to_string);
        pkg.checksum = entry
            .get("checksum")
            .and_then(Value::as_str)
            .map(str::to_string);
        if let Some(deps) = entry.get("dependencies").and_then(Value::as_mapping) {
            for (dep_name, _) in deps {
                if let Some(dep_name) = dep_name.as_str() {
                    pkg.deps.push(Dep::by_name(dep_name));
                }
            }
        }
        lf.push(pkg);
    }

    Ok(lf)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_berry_format() {
        let src = r#"__metadata:
  version: 8
  cacheKey: 10

"express@npm:^4.18.0":
  version: 4.18.2
  resolution: "express@npm:4.18.2"
  dependencies:
    accepts: "npm:~1.3.8"
  checksum: 10c0/abc

"accepts@npm:~1.3.8":
  version: 1.3.8
  resolution: "accepts@npm:1.3.8"
  checksum: 10c0/def
"#;
        let lf = parse(src).unwrap();
        assert_eq!(lf.format_version.as_deref(), Some("8"));
        assert_eq!(lf.packages.len(), 2);
        let express = lf.packages.iter().find(|p| p.name == "express").unwrap();
        assert_eq!(express.version, "4.18.2");
        assert_eq!(express.source.as_deref(), Some("express@npm:4.18.2"));
        assert_eq!(express.checksum.as_deref(), Some("10c0/abc"));
        assert_eq!(express.deps[0].name, "accepts");
    }

    #[test]
    fn dispatches_to_classic_without_metadata() {
        let src = "lodash@^4.17.21:\n  version \"4.17.21\"\n  integrity sha512-x\n";
        let lf = parse(src).unwrap();
        assert_eq!(lf.format_version.as_deref(), Some("1"));
        assert_eq!(lf.packages[0].name, "lodash");
    }
}
